### ISSUE TITLE
BKSF-Mapper

### ISSUE THREAD

**Issue by manoelmarques** [Qiskit/qiskit-nature#9]
Migrated from Enterprise Github.
Creator: @mrossinek

### What is the expected behavior?

Implement and test the `BravyiKitaevSuperFastMapper`.
This issue breaks-out of #18 and can be used to assign and discuss BKSF-related topics.

- [x] Implementation
- [x] Unittest
- [x] Documentation

**Comment by mrossinek** [comment]
I would like to point out the proposed re-design of the mappers in terms of "label-translators" as explained here: https://github.com/Qiskit/qiskit-nature/issues/18#issuecomment-773494151

I am not sure that it is worth opening another issue to propose a refactoring of the mappers but this will likely be the necessary implementation approach for the `BKSF` mapper. When this issue gets tackled we can potentially re-design the existing mappers accordingly.

The benefit of such an implementation would include:
- a clearer and more maintainable implementation (i.e. no "odd" `pauli_table` object)
- an implementation which is inspired closely by the actual mathematical definitions of the mappings
- a more abstract implementation allowing greater flexibility for future development of other mappings

**Comment by jlapeyre** [comment]
I favor retaining a class that stores the fermionic Hamiltonian as 2-d and 4-d arrays, as in aqua's `FermionicOperator`. And furthermore that this class be exposed in the API.

* This is essentially how the operators are produced by the electronic-structure software. Going through a string representation, although it offers some advantages, is more complicated.
* Because it's more complicated, it is susceptible to bugs and performance issues. For example, currently the performance of `FermionicOp` prohibits mapping Hamiltonians that can easily be done using the simpler format. This can be performance can be fixed. But, the representation of Fermionic operators has been in flux and may continue to change in the future.
* The direct representation serves as a control for the correctness of the derived one. This is important because the software is still in a developmental/experimental stage.

Disadvantage
* A `FermionicOperator`-like class, if exposed, increases the API and maintenance burden.

**Comment by mrossinek** [comment]
I do see where you're coming from with this. However, I also see some disadvantages with this approach.
For one, your suggestion inherently limits us to one- and two-body terms whereas the current implementation can easily be generalized to n-body terms (especially with #220 well underway). Although (right now) none of our classical drivers supports n-body terms, this should not be limitation which we restrict ourselves too.

If, however, BKSF by its nature does not support n-body terms anyways, then this is a different discussion all together. Previously,  a suggestion was brought up to provide `to_matrix()` conversion methods on the `SecondQOp`s (#129). I am not sure what the status of that discussion is (@pbark @LaurinFischer @ikkoham), but would that resolve some of the issues you are facing? Or do you explicitly need to retain the information whether you were dealing with 1- or 2-body terms?

In terms of performance: I agree, that things are not ideal right now. And a more general framework is usually less performant than  a "restricted" one. But I also believe that we can still improve the performance significantly (see also #234).

And a final note on inter-operability: It may be true, that this is currently not straight forward, but I don't think it is impossible to do. And since this is not an essential part of the general workflow, this does not necessarily need to be the most performant code either. So a simple set of conversion methods should be sufficient.

**Comment by jlapeyre** [comment]
* I think I wasn't clear enough. I'm not proposing getting rid of the string-based representation in favor of the tensor-based. I was thinking of retaining both. And don't use the latter in cases where it doesn't apply, i.e. more than one- and two-body terms.

* I have no doubt the performance can be improved. The point is that it had degraded and has to be fixed. And there will likely be further bugs or performance issues in the future. They can be fixed, too. Hopefully, they don't make it through to a release, but maybe they will.  The tensor-based representation is simpler and more widely understood.

I'm not 100% convinced that it should be retained, or in what way it should be retained. I'm mainly concerned about throwing away simple, working, relatively performant, structures and routines in favor of something that is more complicated and under development.

**Comment by woodsp-ibm** [comment]
This is related to what John is saying about integrals as input for an operator. Some people have integrals from elsewhere and had been creating the FermionicOperator in the past and are now lost. They don't start from driver. I fielded another such question today. An operator such as John describes might be helpful for those that prefer to experiment more and are coming at the code from different angles. Of course another way, and its what I refer people too, would be to simply expose the integrals builder function that builds out the FermionicOp from one and two body ints.  While its there in the code it does not come out in the published API ref, not do we have it in any init file where we might want to expose it etc. I know this is not exactly related to BKSF but figured I'd mention since this alterate operator construction was being discussed here anyway - more though in view of BKSF of course.

**Comment by jlapeyre** [comment]
I also thought of having something like the old `FermionicOperator` just as a way to organize and expose an API for building `FermionicOp`. I agree with @woodsp-ibm that this could be really useful.

**Comment by mrossinek** [comment]
With #220 merged, you can use the `ElectronicEnergy` to build an `FermionicOp` from integrals like so:
```python
property = ElectronicEnergy(ElectronicBasis.MO, {
    1: OneBodyIntegrals(ElectronicBasis.MO, (qmol.mo_onee_ints, qmol.mo_onee_ints_b),
    2: TwoBodyIntegrals(ElectronicBasis.MO, (qmol.mo_eri_ints, qmol.mo_eri_ints_ba, qmol.mo_eri_ints_bb, None),
})

fermionic_op: List[FermionicOp] = property.second_q_ops()  # length == 1
```
The syntax may improve slightly during the work on #243 but the overall benefit of this approach is its generality because we are able to easily extend this approach to include n-body integrals in the future.

**Comment by mrossinek** [comment]
Closed by #232
